File: mcp/src/tools/edges.ts
import type { YaadIndexClient } from "../client/yaad_index.js";

export const edgesTool = {
 name: "edges",
 description:
 "Single-hop edge query. Wraps " +
 "`GET /v1/edges?entity_id=X[&edge_types=...][&direction=...]`. " +
 "Returns the typed edges from/to the entity in one round-trip. " +
 "Distinct from `get_entity_with_context`: this is the **flat one-hop** " +
 "surface (\"who designed this game\" / \"what cites this person\"); " +
 "`get_entity_with_context` does N-hop BFS traversal — overkill for " +
 "direct-edge queries. " +
 "Direction defaults to **out** (backward-compat with the existing " +
 "per-entity edge-expansion semantic). `direction=in` returns " +
 "inbound edges only (entity_id matches `to_id`); `direction=both` " +
 "returns inbound + outbound combined. " +
 "`edge_types` is an optional allowlist; absent → no type filter " +
 "(every edge type returns). " +
 "Response shape: `{ok, edges: [{type, from_id, to_id, metadata?}], " +
 "next_cursor}`. `next_cursor` is reserved for forward-compat and " +
 "always null today (single-hop counts per entity are bounded).",
 inputSchema: {
 type: "object",
 properties: {
 entity_id: {
 type: "string",
 description:
 "Full entity id in the `<kind>:<local-id>` shape — e.g. " +
 "`boardgame:brass-birmingham-2018`. The role (from vs " +
 "to) depends on direction.",
 },
 edge_types: {
 type: "array",
 items: { type: "string" },
 description:
 "Optional allowlist of edge types to return (e.g. " +
 "`[\"designed_by\", \"authored_by\"]`). Absent or empty " +
 "→ no filter, every edge type returns.",
 },
 direction: {
 type: "string",
 enum: ["out", "in", "both"],
 description:
 "Edge direction relative to entity_id. `out` (default) " +
 "= entity_id is from_id. `in` = entity_id is to_id. " +
 "`both` = inbound + outbound combined.",
 },
 },
 required: ["entity_id"],
 },
} as const;

export async function runEdges(
 client: YaadIndexClient,
 args: Record<string, unknown>,
): Promise<unknown> {
 const id = String(args.entity_id ?? "");
 if (!id) {
 return { ok: false, error: "invalid_argument", message: "`entity_id` is required" };
 }
 let edgeTypes: string[] | undefined;
 if (args.edge_types !== undefined && args.edge_types !== null) {
 if (!Array.isArray(args.edge_types)) {
 return { ok: false, error: "invalid_argument", message: "`edge_types` must be an array of strings" };
 }
 edgeTypes = args.edge_types.map((s) => String(s));
 }
 const direction = args.direction;
 if (direction !== undefined && direction !== "out" && direction !== "in" && direction !== "both") {
 return {
 ok: false,
 error: "invalid_argument",
 message: "`direction` must be one of {out, in, both}",
 };
 }
 return client.edges(id, edgeTypes, direction as "out" | "in" | "both" | undefined);
}

File: internal/mcp/edges.go
// edges tool — single-hop edge query around an entity.

package mcp

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/url"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func registerEdges(s *server.MCPServer, b *bridge) {
	tool := mcp.NewTool("edges",
		mcp.WithDescription(
			"Single-hop edge query. Wraps "+
				"`GET /v1/edges?entity_id=X[&edge_types=...][&direction=...]`. "+
				"Returns the typed edges from/to the entity in one round-trip. "+
				"Distinct from `get_entity_with_context`: this is the flat "+
				"one-hop surface; `get_entity_with_context` does N-hop BFS. "+
				"Direction defaults to `out` (entity_id is from_id); `in` "+
				"returns inbound edges; `both` returns both combined. "+
				"`edge_types` is an optional allowlist.",
		),
		mcp.WithString("entity_id",
			mcp.Required(),
			mcp.Description("Full entity id in `<kind>:<local-id>` shape."),
		),
		mcp.WithArray("edge_types",
			mcp.Description("Optional allowlist of edge types."),
			mcp.Items(map[string]any{"type": "string"}),
		),
		mcp.WithString("direction",
			mcp.Description("Edge direction relative to entity_id. `out` (default) | `in` | `both`."),
			mcp.Enum("out", "in", "both"),
		),
	)
	s.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		id := req.GetString("entity_id", "")
		if id == "" {
			return mcp.NewToolResultError("`entity_id` is required"), nil
		}
		q := url.Values{}
		q.Set("entity_id", id)
		if direction := req.GetString("direction", ""); direction != "" {
			q.Set("direction", direction)
		}
		if types, ok := req.GetArguments()["edge_types"].([]any); ok && len(types) > 0 {
			for _, t := range types {
				if s, ok := t.(string); ok && s != "" {
					q.Add("edge_types", s)
				}
			}
		}
		return b.callTool(ctx, "GET", fmt.Sprintf("/v1/edges?%s", q.Encode()), nil)
	})
}

// registerUpdateEdgeTarget wires the #304 Cut B primitive: a
// single transactional API that rewrites an edge's target. Used
// by agent-driven flows that resolve a name (via plugin search,
// operator-disambiguation task, etc.) and want to swap the
// placeholder target for the resolved canonical entity without
// losing the edge's identity / created_at audit trail.
//
// Returns the new edge envelope on success. Error envelopes
// (passthrough — branch on `ok === false`):
//
//   - `edge_stale` (409) — the (from, type, old_target) tuple
//     doesn't match a current edge. Re-read state + retry with
//     the fresh tuple. Concurrent rewrites converge here.
//   - `missing_entity` (422) — new_target doesn't resolve to a
//     known entity. The caller materializes it first (via
//     `ingest` or by picking a valid canonical id) and retries.
//   - `invalid_argument` (400) — any required field empty, or
//     old_target == new_target (no-op).
func registerUpdateEdgeTarget(s *server.MCPServer, b *bridge) {
	tool := mcp.NewTool("update_edge_target",
		mcp.WithDescription(
			"Rewrite an edge's target in a single transaction. "+
				"Wraps `POST /v1/edges/update-target`. Deletes "+
				"`(from, type, old_target)` and creates "+
				"`(from, type, new_target)` preserving the "+
				"original edge's `created_at` + metadata so the "+
				"audit trail shows \"edge existed since T, target "+
				"finalized at T'\" rather than a delete+create pair. "+
				"Stale-safety: returns `edge_stale` (409) when the "+
				"old tuple doesn't match current state (already "+
				"rewritten, deleted, or never existed) so concurrent "+
				"rewrites converge cleanly. `new_target` MUST "+
				"reference an existing entity; absent → "+
				"`missing_entity` (422). Used by agent flows that "+
				"resolve a disambiguation and need to swap a "+
				"placeholder edge target for the resolved canonical "+
				"id without losing edge identity.",
		),
		mcp.WithString("from",
			mcp.Required(),
			mcp.Description("Source entity id."),
		),
		mcp.WithString("type",
			mcp.Required(),
			mcp.Description("Edge type (must be a registered edge_kind)."),
		),
		mcp.WithString("old_target",
			mcp.Required(),
			mcp.Description("Current target entity id of the edge being rewritten."),
		),
		mcp.WithString("new_target",
			mcp.Required(),
			mcp.Description("Replacement target entity id."),
		),
	)
	s.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		from := req.GetString("from", "")
		if from == "" {
			return mcp.NewToolResultError("`from` is required"), nil
		}
		edgeType := req.GetString("type", "")
		if edgeType == "" {
			return mcp.NewToolResultError("`type` is required"), nil
		}
		oldTarget := req.GetString("old_target", "")
		if oldTarget == "" {
			return mcp.NewToolResultError("`old_target` is required"), nil
		}
		newTarget := req.GetString("new_target", "")
		if newTarget == "" {
			return mcp.NewToolResultError("`new_target` is required"), nil
		}
		body, err := json.Marshal(map[string]any{
			"from":       from,
			"type":       edgeType,
			"old_target": oldTarget,
			"new_target": newTarget,
		})
		if err != nil {
			return mcp.NewToolResultError("encode args: " + err.Error()), nil
		}
		return b.callTool(ctx, "POST", "/v1/edges/update-target", bytes.NewReader(body))
	})
}
